openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,435 @@
+		})
+}
+
+// managementCPUsOverride presents admission plugin that should replace pod container CPU requests with a new management resource.
+// It applies to all pods that:
+// 1. are in an allowed namespace
+// 2. and have the workload annotation.
+//
+// It also sets the new management resource request and limit and  set resource annotation that CRI-O can
+// recognize and apply the relevant changes.
+// For more information, see - https://github.com/openshift/enhancements/pull/703
+//
+// Conditions for CPUs requests deletion:
+// 1. The namespace should have allowed annotation "workload.openshift.io/allowed": "management"
+// 2. The pod should have management annotation: "workload.openshift.io/management": "{"effect": "PreferredDuringScheduling"}"
+// 3. All nodes under the cluster should have new management resource - "management.workload.openshift.io/cores"
+// 4. The CPU request deletion will not change the pod QoS class
+type managementCPUsOverride struct {
+	*admission.Handler
+	nsLister       corev1listers.NamespaceLister
+	nsListerSynced func() bool
+	nodeLister     corev1listers.NodeLister
+	nodeListSynced func() bool
+}
+
+func (a *managementCPUsOverride) SetExternalKubeInformerFactory(kubeInformers informers.SharedInformerFactory) {
+	a.nsLister = kubeInformers.Core().V1().Namespaces().Lister()
+	a.nsListerSynced = kubeInformers.Core().V1().Namespaces().Informer().HasSynced
+	a.nodeLister = kubeInformers.Core().V1().Nodes().Lister()
+	a.nodeListSynced = kubeInformers.Core().V1().Nodes().Informer().HasSynced
+}
+
+func (a *managementCPUsOverride) ValidateInitialization() error {
+	if a.nsLister == nil {
+		return fmt.Errorf("%s did not get a namespace lister", PluginName)
+	}
+	if a.nsListerSynced == nil {
+		return fmt.Errorf("%s plugin needs a namespace lister synced", PluginName)
+	}
+	if a.nodeLister == nil {
+		return fmt.Errorf("%s did not get a node lister", PluginName)
+	}
+	if a.nodeListSynced == nil {
+		return fmt.Errorf("%s plugin needs a node lister synced", PluginName)
+	}
+	return nil
+}
+
+func (a *managementCPUsOverride) Admit(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	workloadAnnotations := getWorkloadAnnotations(pod.Annotations)
+
+	// strip any resource annotations specified by a user
+	stripResourcesAnnotations(pod.Annotations)
+	// strip any workload annotation to prevent from underlying components(CRI-O, kubelet) to apply any changes
+	// according to the workload annotation
+	stripWorkloadAnnotations(pod.Annotations)
+
+	// more than one workload annotation exists under the pod and we do not support different workload types
+	// under the same pod
+	if len(workloadAnnotations) > 1 {
```

Comment:
We only allow exactly one and we know which one today, right?

Reply:
There's nothing particularly special about the `management` name, except that we're using it for core workloads in the first use case, so we're trying not to hard-code it. There is an idea that as part of making this feature future-proof and flexible, we would allow multiple annotations targeting different workload types with some form of precedence to let the system choose between them based on how it is configured.